feat: Allow multiple role profile for user

niraj2477 opened the pull request:
For every user, setting up the role profile helps to group their role. However, as the Role Profile has the link field, it allows only one profile per user. In case of users where they can have multiple role profile, one has to duplicate all the role from multiple profile to create new role profile.

## What's changed:
-  Instead of the link field for `Role Profile`, add `Table MultiSelect`. Which will allow Admins to allocate multiple `Profiles` to the particular user.
- For example:  `Employee` Profile which will be allocated to everyone in the organization, and then the `Project Manager` Profile can be allocated to only managers.
- Patch for existing `Role Profile` has been added.


closes https://github.com/frappe/frappe/issues/24486

niraj2477 committed:
feat: Add child doctype for storing role profile of users

niraj2477 committed:
fix: Update the user method to store multiple role according to role profile

niraj2477 committed:
feat(patch): Update existing role profiles

niraj2477 committed:
fix(patch): Update patch file

niraj2477 committed:
fix: Remove debug statement

niraj2477 committed:
fix: Resolve lint issue

niraj2477 committed:
Merge branch 'frappe:develop' into multiple-role-profile

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -214,9 +214,9 @@
    "allow_in_quick_entry": 1,
    "fieldname": "role_profile_name",
-   "fieldtype": "Link",
+   "fieldtype": "Table MultiSelect",
```
We shouldn't reuse same field. It also breaks semantics. It should be called `role_profiles`. 

If we add a new field then can safely make it backward compatible too.

ankush reviewed (changes requested):
@niraj2477 few comments from first cut review:
-  can you setup pre-commit https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist#linting we have disable bypassing of branch protection rules for _everyone_, so these have to be fixed :smile: 
- This is critical change from permissions POV, it should have multiple tests for validating it works. 
    - Fix/change current single role profile tests and fixtures. 
    - Add test for 2-3 role profiles being used at same time.
    - Add test for updating role profiles. This will fail as it doesn't seem to be handled in PR.  Refer: https://github.com/frappe/frappe/blob/6ef78c28645b9f1550dd8958770835daf03bd2d2/frappe/core/doctype/role_profile/role_profile.py#L34-L56

niraj2477 commented:
>can you setup pre-commit https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist#linting 

When I ran pre-commit, I noticed that the Doc string was being converted to `Tabs`, which was `Spaces` before, due to which the Doc String was shifted too much 😅.

niraj2477 committed:
fix(Schema): Update the field name in User schema

niraj2477 committed:
fix: Update fields names for whitlist method

niraj2477 committed:
fix: Update method for updating role in role profiles for user

niraj2477 committed:
fix(test): pass list for role profile in existing  single role profile test

niraj2477 committed:
feat(test): Add test case for multiple role profile assignments

niraj2477 committed:
fix(patch): Update field name

niraj2477 commented:
`Semgrep` check can be addressed in separate PR 🤔

ankush commented on frappe/core/doctype/role_profile/role_profile.py:
```diff
@@ -35,12 +35,14 @@ def update_all_users(self):
-			.where(user.role_profile_name == self.name)
+			.join(user_role_profile)
+			.on(user.name == user_role_profile.parent)
+			.where(user_role_profile.role_profile == self.name)
```
Rest of this function body also needs to be changed. That logic doesn't make sense anymore. 


Currently, on role profile update we see if all users with this role profile have exact same roles. This optimization however is doesn't apply when their roles are mixture of multiple role profiles. Maybe we'll have to think of some better way because sites often have 1000s of users with same role profile and updating them if they are not impacted is quite costly.

niraj2477 replied:
1. Fetch users which has current role profile assigned to them. 
2. Make a set of previous role profile's roles and updated one, and re-assign the roles.

This was we can limit the users we are updating. 🤔  @ankush

ankush commented on frappe/core/doctype/user/user.js:
```diff
@@ -27,12 +27,12 @@ frappe.ui.form.on("User", {
 
-	role_profile_name: function (frm) {
-		if (frm.doc.role_profile_name) {
+	role_profiles: function (frm) {
```
Does this work on tables? :thinking: (not sure honestly)

niraj2477 replied:
Yes.

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -211,14 +211,6 @@
    "permlevel": 1,
    "read_only": 1
   },
-  {
-   "allow_in_quick_entry": 1,
-   "fieldname": "role_profile_name",
-   "fieldtype": "Link",
-   "label": "Role Profile",
-   "options": "Role Profile",
-   "permlevel": 1
-  },
```
Let's not remove this right now. Better solution for backward compatibility:

1. Add new field.
2. On save old field is moved to new filed automatically (also done via patch)
3. Hide the old field and we remove it completely in v16. 


If we don't do this it will break all current code that programmatically assigns role profiles.

niraj2477 replied:
> On save old field is moved to new filed automatically (also done via patch) 

This is done via patch.

ankush replied:
We should still do it on save because existing code doesn't know anything about this table.

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -700,6 +692,14 @@
    "label": "Onboarding Status"
+  },
+  {
+   "allow_in_quick_entry": 1,
```
We can remove old one from quick entry now.

ankush commented on frappe/patches/v15_0/migrate_role_profile_to_table_multi_select.py:
```diff
@@ -0,0 +1,13 @@
+		user_role_profile = [{"role_profile": user.role_profile_name}]
+		user = frappe.get_doc("User", user.name)
+		user.update({"role_profiles": user_role_profile})
+		user.save()
```
1. This will be super slow with a large number of users. Let's just insert a child table row for the role profile and not resave the parent document.
2. This will either require a periodic commit or `bulk_insert` to avoid a massive long-running DB transaction.

niraj2477 replied:
1. Fetch `name` for user which has role profile name is set in batch of 10k.
2. Bulk insert the child table data using `bulk_insert` for above user.

@ankush Let me know your thoughts on this, if you have better approach for this.

ankush replied:
@niraj2477 bulk insert is simpler IMO. 

https://github.com/frappe/frappe/blob/1b8e9c3502b702178f777cf2e5c253ccec573e2e/frappe/geo/doctype/country/country.py#L31-L34

niraj2477 committed:
fix: Restore role profile name and hide it

ankush commented:
@niraj2477 rebase branch for new `ruff` formatter, should fix that whitespace issue.

niraj2477 committed:
fix(patch): Use bulk insert child records

niraj2477 committed:
Merge branch 'develop' into multiple-role-profile

niraj2477 committed:
feat: Move old role profile to table multiselect

niraj2477 committed:
Merge branch 'develop' into multiple-role-profile

niraj2477 committed:
fix: Move old field to table multiselect

niraj2477 committed:
fix: Modify function to update role of user on role profile change

niraj2477 committed:
fix: Resolve lint issue

niraj2477 commented:
In connection tab for `Role Profile` link to the `User` will not work as expected as in User DocType it will be MultiSelect. It will show legacy connection.

ankush commented on frappe/core/doctype/role_profile/role_profile.py:
```diff
@@ -34,24 +34,16 @@ def on_update(self):
+		users = frappe.get_list("User Role Profile", filters={"role_profile": self.name}, pluck="parent")
+		for user in users:
+			role_profile_roles = []
+			user = frappe.get_doc("User", user)
```
This will be quite slow :sweat_smile: not sure about alternate implementation yet but this is gonna become a problem some day for sure.

ankush replied:
Optimization: 
- Get users with this role profile, also get all of their role profiles
- Reconstruct new roles that they'll have. If they are same then odn't reupate, else update.
- No need to write any business logic here. user.py is already doing it on save?

niraj2477 replied:
We can optimize it further as you suggested, at least this way we will have advantage in comparison with join queries. 😅 

> No need to write any business logic here. user.py is already doing it on save?

Previously also user.py did not have the logic to update the user, if profile is itself update.  🤔

ankush replied:
@niraj2477 previous implementation avoided saving user if there was no change, here we are re-saving anyway so everything will get triggered. 

https://github.com/frappe/frappe/pull/24745/commits/9c2c3112a036cdb7ae1c2f56b2bb662d745ee37b

ankush commented on frappe/core/doctype/role_profile/role_profile.py:
```diff
@@ -34,24 +34,16 @@ def on_update(self):
+		for user in users:
+			role_profile_roles = []
+			user = frappe.get_doc("User", user)
+			for role_profile in user.role_profiles:
+				if self.name == role_profile.role_profile:
+					continue
+				profile = frappe.get_doc("Role Profile", role_profile.role_profile)
+				role_profile_roles.extend([role.role for role in profile.roles])
+			role_profile_roles.extend([role.role for role in self.roles])
```
umm, why 2 different implementation of same business logic. Previously it was still straightforward, now with multiple profiles it's best to define this in one place only.

niraj2477 replied:
If the role profile itself is updated, then we will have to update the users also.

ankush commented on frappe/core/doctype/user/user.json:
```diff
@@ -212,13 +213,14 @@
-   "options": "Role Profile",
-   "permlevel": 1
-  },
+    "allow_in_quick_entry": 0,
```
Avoid manually editing these files. :sweat_smile:

ankush committed:
fix: disable role editor if role profiles are present

ankush commented on frappe/core/doctype/user/user.py:
```diff
@@ -167,23 +171,47 @@ def validate(self):
+			(profile.role_profile == self.role_profile_name) for profile in self.role_profiles
+		):
+			return
+		profile_exists = frappe.db.exists(
+			"User Role Profile",
+			{"parent": self.name, "role_profile": self.role_profile_name, "parenttype": "User"},
+		)
+		if not profile_exists:
+			user_role_profile = frappe.get_doc(
+				{
+					"doctype": "User Role Profile",
+					"role_profile": self.role_profile_name,
+					"parent": self.name,
+					"parenttype": "User",
+					"parentfield": "role_profiles",
+				}
+			)
+			user_role_profile.save(ignore_permissions=True)
+			self.role_profiles.append(user_role_profile)
```
This is needlessly complicated, you have access to child table, just verify it from there directly.

ankush committed:
refactor: use get_all instead of get_list

There won't be any perm checks here.

ankush committed:
refactor: backward compatiblity code for role_profile_name

ankush committed:
perf: use cached doc for fetching role profiles

ankush committed:
Merge branch 'develop' into multiple-role-profile

ankush commented:
@niraj2477 I am pushing some changes, some new/old code needs changes.

ankush committed:
refactor: avoid implementing duplicate logic for role profile

Unless we are getting any performance benefit, I don't see the need to
have two different implementations for it.

also, queue job in long queue to reduce chance of failure

ankush commented on frappe/core/doctype/role_profile/role_profile.py:
```diff
@@ -34,24 +34,7 @@ def on_update(self):
-				user = frappe.get_doc("User", user)
-				user.roles = []
-				user.add_roles(*role_profile_roles)
+		users = frappe.get_list("User Role Profile", filters={"role_profile": self.name}, pluck="parent")
```
FYI get_list shouldn't be used like this. This is internal code so no perm checks are required. In fact perm checks here is a bug.

niraj2477 replied:
Noted.

ankush committed:
fix: roles getting re-inserted on every save

ankush committed:
fix: deduplicate role profiles

ankush committed:
refactor: remove code duplication for role profile logic

ankush committed:
refactor: useless use of dict in frappe.get_doc

ankush committed:
fix: role profile > user dashboard links

ankush commented:
> In connection tab for Role Profile link to the User will not work as expected as in User DocType it will be MultiSelect. It will show legacy connection.

Fixed, you can link a field in child table. It's quite confusing though.  :sweat_smile:

ankush committed:
test: role ordering doesn't matter
